### PR TITLE
add design doc: xgboost on sqlflow

### PR DESCRIPTION
**Pull request by sperlingxx**

### CODE REVIEW

File: doc/xgboost_on_sqlflow.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+# xgboost on sqlflow
```
Comment (typhoonzero) [inline doc/xgboost_on_sqlflow.md]: Capitalize the first letter in the titles.

Comment (typhoonzero) [reply doc/xgboost_on_sqlflow.md]: This is a design doc, the title and file name should be like `# Design: ....` so that readers won't think it's a tutorial or manual.

File: doc/xgboost_on_sqlflow.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    def write_lines(self, lines: List[str], model_path: str):
+        pass
+``` 
+
+
+With _xgblauncher_, we can launch _xgboost_ from sqlflow AST via a lightweight python `code template` and a corrsponding `filler`.
+The `code template` roughly includes  components as follows: 
+
+* `TFDataSource` that is responsible for fetching and pre-processing data via tf.feature_columns API.
+   Data will be fetched in mini-batch style by executing TF compute graph with mini-batch data feed by `sqlflow_submitter.db.db_generator`.
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md:128]: I am not sure if XgBoost is compatible with `tf.feature_column`. What kind of feature columns are you planning to support?

File: doc/xgboost_on_sqlflow.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+## xgboost on sqlflow via ant-xgboost
+   
+### Overview
+
+We use [ant-xgboost](https://github.com/alipay/ant-xgboost) as backend,
+which is forked from [xgboost](https://github.com/dmlc/xgboost).
+_Ant-xgboost_ is nearly same as original xgboost, 
+except some improvements to make xgboost easier to use, 
+such as better early stopping strategy, parameter checking, and end to end launcher.
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: Please add the documentation links of these features. Maybe also note how these features will benefit SQLFlow users.

File: doc/xgboost_on_sqlflow.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    c1, c2, c3, c4, c5 as class
+from kaggle_credit_fraud_training_data
+TRAIN XGBoostEstimator
+WITH
+  booster = "gbtree"
+  objective = "logistic:binary"
+  eval_metric = "auc"
+  train_eval_ratio = 0.8
+COLUMN
+  NUMERIC(c1) as fc1,
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: SQLFlow hasn't supported `AS` keyword yet.

File: doc/xgboost_on_sqlflow.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: I am no expert in XgBoost. Is it common to bucketize input when training a tree model? I feel a tree model is able to learn the decision boundary.

File: doc/xgboost_on_sqlflow.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  CROSS([fc1, fc2, c3]) as fc3,
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: Similarly for `CROSS` feature. Looks like cross feature is equivalent to a multi-level tree, I feel a tree model is able to learn the cross.

File: doc/xgboost_on_sqlflow.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+LABEL class
+INTO sqlflow_models.xgboost_model_table;
+
+// sample clause of predict
+select 
+    c1, c2, c3, c4
+from kaggle_credit_fraud_development_data
+PREDICT kaggle_credit_fraud_development_data.class
+USING sqlflow_models.xgboost_model_table
+COLUMN
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: There is no `COLUMN` clause in prediction.

File: doc/xgboost_on_sqlflow.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  c4 as fc4;
+```
+
+### Implementation
+
+As `codegen.go` generating _TensorFlow_ code from sqlflow AST,
+we will add `codegen_xgboost.go` which translate sqlflow AST into a python launcher program of _xgboost_. 
+
+In _ant-xgboost_, there exists an incubating module named [_xgblauncher_](https://github.com/alipay/ant-xgboost/tree/ant_master/xgboost-launcher), 
+an extendable, cloud-native xgboost based machine learning pipeline,
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: Please add the documentation links of "an extendable, cloud-native XgBoost based machine learning pipeline".

File: doc/xgboost_on_sqlflow.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+   Data will be fetched in mini-batch style by executing TF compute graph with mini-batch data feed by `sqlflow_submitter.db.db_generator`.
+
+* `DBDataSource` that is responsible for writing prediction results into specific data base.
+   The writing action can be implemented via `sqlflow_submitter.db.insert_values`.
+
+* `LocalModelSource` that is responsible for reading/writing _xgboost_ models on local file system.
+
+* Configuration conversions and entry point of _xgblauncher_.
+
+
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md:150]: Maybe add a "Distributed training" section.

File: doc/xgboost_on_sqlflow.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
+# xgboost on sqlflow
+
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md:2]: Add a few sentences summarizing the content of this doc. [ref](https://www.freecodecamp.org/news/how-to-write-a-good-software-design-document-66fcf019569c/)

> A high-level summary that every engineer at the company should understand and use to decide if it’s useful for them to read the rest of the doc. It should be 3 paragraphs max.

File: doc/xgboost_on_sqlflow.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+## XGBoost
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md]: Please be consistent on the naming: `xgboost`, `XGBoost` or `XgBoost`?

File: doc/xgboost_on_sqlflow.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    @abstractmethod
+    def read(self) -> Iterator[XGBoostRecord]:
+        pass
+
+    @abstractmethod
+    def write(self, result_iter: Iterator[XGBoostResult]):
+        pass
+
+    
+class ModelSource:
```
Comment (tonyyang-svail) [inline doc/xgboost_on_sqlflow.md:98]: How is this class different from XgBoost's `save/load`? [ref](https://xgboost.readthedocs.io/en/latest/python/python_intro.html#training)

**Review by tonyyang-svail — COMMENTED** [review]
@sperlingxx Thanks for submitting this PR. Commented.

File: doc/xgboost_on_sqlflow.md (hunk 2 above)
Comment (sperlingxx) [reply doc/xgboost_on_sqlflow.md:128]: We want to support as many `feature_columns` as sqlflow semantic does.
Although feeding xgboost with vectorized features like bucketized vectors is not always a good idea, xgboost can handle them anyway.  Since linear model is available in xgboost as an alternative to tree model,  I think it is reasonable to train xgboost models with vectorized features.

File: doc/xgboost_on_sqlflow.md (hunk 12 above)
Comment (sperlingxx) [reply doc/xgboost_on_sqlflow.md:98]: `ModelSource`is a wrapper of `booster.save/load`,  it integrate saving work of booster binary, booster info and  feature scores.  The purpose of `MdoelSource` is providing an interface to save/load xgboost models on different stores. 
Here is the full implemention:  https://github.com/alipay/ant-xgboost/blob/ant_master/xgboost-launcher/launcher/model_source.py

File: doc/xgboost_on_sqlflow.md (hunk 5 above)
Comment (yinlou) [reply doc/xgboost_on_sqlflow.md]: Bucketization helps the training process; it's a standard technique used by almost all boosted tree implementations. The question here is whether a user should explicitly do bucketization. I know some cases where bucketization is done on user side so this actually looks okay to me. If you think it adds confusion, we can take it off from this example?

File: doc/xgboost_on_sqlflow.md (hunk 6 above)
Comment (yinlou) [reply doc/xgboost_on_sqlflow.md]: Agreed tree can automatically learn feature interaction, although empirical study often finds adding human-crafted interactions can further improve accuracy.

File: doc/xgboost_on_sqlflow.md (hunk 8 above)
Comment (sperlingxx) [reply doc/xgboost_on_sqlflow.md]: Doc of xgblauncher will be available in antxgboost soon.

File: doc/xgboost_on_sqlflow.md (hunk 12 above)
Comment (tonyyang-svail) [reply doc/xgboost_on_sqlflow.md:98]: I am curious why XgBoost's `save` doesn't save the booster info...

File: doc/xgboost_on_sqlflow.md (hunk 2 above)
Comment (tonyyang-svail) [reply doc/xgboost_on_sqlflow.md:128]: Sure, you can find the list of currently supported feature column here: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/user_guide.md#feature-columns.

File: doc/xgboost_on_sqlflow.md (hunk 5 above)
Comment (tonyyang-svail) [reply doc/xgboost_on_sqlflow.md]: Great. Let's keep it in the example since `BUCKETIZED` is commonly used.

File: doc/xgboost_on_sqlflow.md (hunk 6 above)
Comment (tonyyang-svail) [reply doc/xgboost_on_sqlflow.md]: Cool, let's take `CROSS` off since AntXgBoost is aiming for full AutoML.

File: doc/xgboost_on_sqlflow.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+which is forked from [xgboost](https://github.com/dmlc/xgboost).
+_Ant-xgboost_ is nearly same as original xgboost, 
+except some improvements to make xgboost easier to use, 
+such as better early stopping strategy, parameter checking, and end to end launcher.
+
+### User Experience
+    
+In terms of sqlflow users, _xgboost_ is an alternative _Estimator_ like _Tensorflow Estimators_. 
+Working with xgboost is quite similar to working with Tensorflow estimators; 
+just change `TRAIN DNNClassifier` into `TRAIN XGBoostEstimator`.
```
Comment (Yancey0623) [inline doc/xgboost_on_sqlflow.md]: Currently, SQLFlow using the env `SQLFlow_submittor` to support the different submitters(`alps` for ALPS lib or Tensorflow Keras by default): 

https://github.com/sql-machine-learning/sqlflow/blob/317ef496f36f5412f47d9f56ba894d590844bfa4/sql/executor.go#L223-L225

Maybe we can follow this logic instead of using the prefix of the model name. In addition, SQLFlow will support the different backend by passing  the `submitter` field in the Session struct: https://github.com/sql-machine-learning/sqlflow/issues/539

File: doc/xgboost_on_sqlflow.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+except some improvements to make xgboost easier to use, 
+such as better early stopping strategy, parameter checking, and end to end launcher.
+
+### User Experience
+    
+In terms of sqlflow users, _xgboost_ is an alternative _Estimator_ like _Tensorflow Estimators_. 
+Working with xgboost is quite similar to working with Tensorflow estimators; 
+just change `TRAIN DNNClassifier` into `TRAIN XGBoostEstimator`. 
+In addition, some xgboost specific parameters are required, 
+which can be configured in the same way as _Tensorflow_ parameters.
```
Comment (Yancey0623) [inline doc/xgboost_on_sqlflow.md]: _Tensorflow_  => TensorFlow

File: doc/xgboost_on_sqlflow.md (hunk 7 above)
Comment (Yancey0623) [inline doc/xgboost_on_sqlflow.md]: We can save the column info while saving the model so that don't need to specify the COLUMN repeatedly in the PREDICT statement.

File: doc/xgboost_on_sqlflow.md (hunk 9 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow.md:150]: @sperlingxx , I found [xgboost operator](https://github.com/kubeflow/xgboost-operator), do you have any experience on it? Or maybe you have some best practices about distributed xgboost?

File: doc/xgboost_on_sqlflow.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+For now, _xgboost_ has been bind to some of most popular distributed computing frameworks, such as _Apache Spark_, _Apache Flink_, _Dask_.
+However, specific computing frameworks are not always available in production environments. 
+So, we propose a cloud-native approach: running _xgboost_ directly on _k8s_ cluster. 
+ 
+As _xgblauncher_ is scalable and docker-friendly, xgblauncher-based containers can be easily orchestrated by [xgboost operator](https://github.com/kubeflow/xgboost-operator),
+a specific kubernetes controller for (distributed) xgboost jobs.
+With the help of `xgboost operator`, it is easy to handle `XGBoostJob` via `kuberentes API`, a kubernetes' custom resource defined by `xgboost operator`. 
+
+`XGBoostJob` building and tracking will be integrated to _xgblauncher_ in near future. 
+After that, we can generate python codes with an option to decide whether running xgboost job locally or submitting it to a remote k8s cluster.
```
Comment (Yancey0623) [inline doc/xgboost_on_sqlflow.md]: Sounds great!

File: doc/xgboost_on_sqlflow.md (hunk 10 above)
Comment (sperlingxx) [reply doc/xgboost_on_sqlflow.md:2]: @tonyyang-svail  I have reorganized the doc, splitting it into _overview_, _context_ and _proposed solution_.

**Review by tonyyang-svail — APPROVED** [review]
Thanks @sperlingxx for this excellent design doc. LGTM++.

Feel free to update this design during the implementation. :)

**Review by Yancey0623 — APPROVED** [review]
LGTM

File: doc/xgboost_on_sqlflow.md (hunk 9 above)
Comment (merlintang) [reply doc/xgboost_on_sqlflow.md:150]: @Yancey1989 xgboost operator is good for distributed xgboost training and batch prediction. xgboost operator would help sqlflow-xgboost distributed xgboost training on k8s.
